### PR TITLE
Overridable safety

### PR DESCRIPTION
**Pull request by bcoca**
cause people want to do bad things ...
##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

inventory plugins

### CODE REVIEW

**Review by s-hertel — APPROVED** [review]

**Comment by AlanCoding** [comment]
Link https://github.com/ansible/ansible/issues/40581

This does not resolve that issue, but it would be used to resolve that issue with a followup patch for ec2. Similar patches would be needed for the other plugins.

**Comment by AlanCoding** [comment]
first round of testing:

```
ANSIBLE_TRANSFORM_INVALID_GROUP_CHARS=false ansible-inventory -i private/ec2/aws_ec2.yaml --list --export -vvv
ansible-inventory 2.8.0.dev0
  config file = None
  configured module search path = [u'/Users/alancoding/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/alancoding/Documents/repos/ansible/lib/ansible
  executable location = /Users/alancoding/.virtualenvs/ansible/bin/ansible-inventory
  python version = 2.7.11 (default, Oct 17 2016, 14:59:40) [GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
No config file found; using defaults
host_list declined parsing /Users/alancoding/Documents/repos/ansible-inventory-file-examples/private/ec2/aws_ec2.yaml as it did not pass it's verify_file() method
script declined parsing /Users/alancoding/Documents/repos/ansible-inventory-file-examples/private/ec2/aws_ec2.yaml as it did not pass it's verify_file() method
toml declined parsing /Users/alancoding/Documents/repos/ansible-inventory-file-examples/private/ec2/aws_ec2.yaml as it did not pass it's verify_file() method
 [WARNING]:  * Failed to parse /Users/alancoding/Documents/repos/ansible-inventory-file-examples/private/ec2/aws_ec2.yaml with auto plugin: global name 'get_option' is not defined

  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 272, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/auto.py", line 55, in parse
    plugin.parse(inventory, loader, path, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/aws_ec2.py", line 531, in parse
    if get_option('use_legacy_script_group_name_sanitization'):
```

**Comment by AlanCoding** [comment]
To update, the current effort will very much resolve #40581, the way it is being done will create the need for a follow-on patch for azure_rm (not filed separately), and then gcp_compute (https://github.com/ansible/ansible/issues/53460)

File: lib/ansible/plugins/inventory/aws_ec2.py [hunk 1]
```diff
@@ -520,6 +528,9 @@ def verify_file(self, path):
     def parse(self, inventory, loader, path, cache=True):
         super(InventoryModule, self).parse(inventory, loader, path)
 
+        if get_option('use_legacy_script_group_name_sanitization'):
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/aws_ec2.py]: Along with changing to self.get_option, this should move below _read_config_data (which is adding the options in the config file with set_options). Also need to import re, then this works for me.

**Comment by AlanCoding** [comment]
I want to be clear about what this doesn't do. It does not restore the old behavior of the contrib script `ec2.py`. I will illustrate with some snippets from its output, when run through `ansible-inventory`:

### ec2.py with replace_dash_in_groups = True

(default behavior)

```json
"us-east-1": {
    "children": [
        "us-east-1a",
```

```json
"tag_Test_Flag_2202_Replace_Dash_In_Groups": {
    "hosts": [
        "100.24.5.31", 
        "100.26.193.109",
```

### ec2.py with replace_dash_in_groups = False

```json
"us-east-1": {
    "children": [
        "us-east-1a", 
```

```json
"tag_Test-Flag-2202_Replace-Dash-In-Groups": {
    "hosts": [
        "107.21.83.208", 
        "107.22.128.60", 
```

I'm still thinking through the implications of this.

I might go out on a limb and say that we need https://github.com/ansible/ansible/pull/53446, in addition to this.

To restore the behavior of the script, we would make the groups based off of `placement.region` and `placement.availability_zone` to _always_ use the unsafe names - because this is what the script did. This is true for the default, and true for the case that the user indicates that dashes should be replaced with underscores.

**Comment by s-hertel** [comment]
Currently you can use replace for compatibility with replace_dash_in_groups. The problem before was turning that *off* - there's been no issue replacing hyphens with underscores, just the other way around.

Another option is adding another option to aws_ec2 that can be combined with use_legacy_script_group_name_sanitization to change the regex.

**Comment by bcoca** [comment]
@AlanCoding see the regex I implemented, it keeps '-' as valid

**Review by s-hertel — APPROVED** [review]

**Comment by AlanCoding** [comment]
@bcoca 

> see the regex I implemented, it keeps '-' as valid

I need '-' valid in region names, and not valid for other group types. See:

```json
"us-east-1": {
    "children": [
        "us-east-1a",
```

```json
"tag_Test_Flag_2202_Replace_Dash_In_Groups": {
    "hosts": [
```

This is the default ec2.py behavior.

File: changelogs/fragments/allow_bad_things.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+    - Ebmed an overridable static santiization method into base inventory plugin class to allow individual plugins to override
```
Comment (goneri) [inline changelogs/fragments/allow_bad_things.yml]: -> Embed an overridable static sanitization method into base inventory plugin class to allow individual plugins to override

**Comment by s-hertel** [comment]
@AlanCoding This doesn't change the script default behavior. For the plugin see my comment. None of this is the default for the plugin, you have to turn it on and you can make it compatible just like all the other things in your version of https://gist.github.com/s-hertel/089c613914c051f443b53ece6995cc77.

**Comment by AlanCoding** [comment]
I ran

```
ANSIBLE_TRANSFORM_INVALID_GROUP_CHARS=false ansible-inventory -i private/ec2/aws_ec2.yaml --list --export -vvv
```

I got a lot of output like

> [DEPRECATION WARNING]: Ignoring invalid character(s) "set([u'-'])" in group name (ami-293483957). This feature will be removed in version 2.12. Deprecation warnings can be disabled 
by setting deprecation_warnings=False in ansible.cfg.

Then it produced names like `us-east-1a` and `vpc_id_vpc-38472958`.

This does resolve the linked issue. I would still like https://github.com/ansible/ansible/pull/53446 so that I can obtain the default ec2.py behavior, which I described in prior comments.

**Comment by s-hertel** [comment]
Using a simple replace or another toggle to pair with use_legacy_script_group_name_sanitization is going to be a more compatible solution with the script than #53446. With #53446 you can create groups with ':' and other illegal characters.

Another toggle would be more convenient, but to do there would need to be a whitelist of the variables affected. That whitelist would include things like the ami id, instance type, and tags, really, anything that calls to_safe in the contrib script is affected by that toggle, such as these... https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L983,
https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L990
https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L997
https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L1005
https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L1014
https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L1021
https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L1030

Not sure which is better: hard-code that into aws_ec2 or just use the replace in your YAML file for the affected groups.

**Comment by AlanCoding** [comment]
> Not sure which is better: hard-code that into aws_ec2 or just use the replace in your YAML file for the affected groups.

I'm sorry, I'm not following what is being suggested for either of these options right now. The PR https://github.com/ansible/ansible/pull/53446 is not touching awx_ec2, so I don't know what you mean when you say "hard-code that into aws_ec2".

For the second suggestion, I think you mean the `.replace()` python method. I'd expect you mean something like this:

```yaml
  - prefix: ""
    separator: ""
    key: placement.region.replace('_', '-')
    parent_group: regions
  - prefix: ""
    separator: ""
    key: placement.availability_zone.replace('_', '-')
    parent_group: zones
```

That does not work, for the simple reason that the sanitization is done after the templating. It gives group names like `us_east_1a`. If you could suggest to me any other syntax, I will be happy to test it using the code in the current state of this PR.

**Comment by AlanCoding** [comment]
Also, so that I'm clear how I would do it with https://github.com/ansible/ansible/pull/53446, the translation layer would look like this:

```yaml
  - prefix: ""
    separator: ""
    key: placement.region
    parent_group: regions
    unsafe: true
  - prefix: ""
    separator: ""
    key: placement.availability_zone
    parent_group: zones
    unsafe: true
```

We don't need to entertain hypotheticals about other unsafe characters that might be allowed, because the group names from these two things are always of the general pattern "xx-xxxx-xx".

**Comment by s-hertel** [comment]
@AlanCoding `regex_replace` and `replace` are just different syntax. The tags key should use | regex_replace because you're operating on a dict. For hard-coding the whitelist into aws_ec2, that is the alternative solution I'm proposing to regex_replace or replace. Nothing to do with 53446. Try creating a host with unsafe tags (beyond using hyphens) to see what I mean about #53446 not matching script behavior. Not sure how to explain this in a clearer way, I'd recommend just trying it.

**Comment by AlanCoding** [comment]
> @AlanCoding regex_replace and replace are just different syntax. The tags key should use | regex_replace because you're operating on a dict.

Could you modify the YAML I posted in some way to show what you're talking about? The replace syntax I used came from this comment in the original issue:

https://github.com/ansible/ansible/issues/40581#issuecomment-391217093

The reason it doesn't work now is the same reason it didn't work then.

> Try creating a host with unsafe tags (beyond using hyphens) to see what I mean about #53446 not matching script behavior.

I have no intention of using this on tags, this is only used on regions and zones. The reason is that the script ec2.py treated regions and zones differently.

In ec2.py, it doesn't matter whether you set `replace_dash_in_groups` to True or False, it still gives region names with dashes in them. Whatever method you have of reproducing that behavior is fine with me, just let me know and I'll use it. I have only seen one proposal work for that, and that is #53446 with https://github.com/ansible/ansible/pull/53458#issuecomment-470663197

**Comment by s-hertel** [comment]
@AlanCoding This toggle is for allowing hyphens without other bad characters. This matches the replace_dash_in_groups == False behavior that was not before possible with the plugin. You can't use replace/regex_replace to *add* hyphens, nor has anyone claimed that (I'm not sure we'd want to add hyphens as opposed to preserving them).

You can however, replace hyphens with safe characters (like '_') as replace_dash_in_groups == True does for certain groups. I'm suggested adding an alternative toggle such as `use_legacy_replace_dash_in_groups` which would modify the regex to replace '-' for a whitelist of variable names to make replace_dash_in_groups == True more convenient.

I made an example of replace_dash_in_groups = True and replace_dash_in_groups = False. The former would be simplified by my proposed toggle.
https://gist.github.com/s-hertel/e66141d54f4a0c9ed5cc81116dee4428

Anyway, we're going in circles but it looks like https://github.com/ansible/ansible/pull/53435 will test the parity for the broad use cases so it's okay if you don't want to experiment with that. :-)

**Comment by AlanCoding** [comment]
@s-hertel your suggested syntax of the form

```yaml
keyed_groups:
  - prefix: tag
    key: tags | regex_replace('-', '_')
  - prefix: key 
    key: key_name.replace('-', '_')
  - prefix: ""
    separator: ""
    key: placement.region
  - prefix: ""
    separator: ""
    key: placement.availability_zone
  - key: platform.replace('-', '_')
  - key: vpc_id.replace('-', '_')
```

is a very helpful suggestion, where the outlook was previously rather bleak. This appears as a valid alternative where I was coming up short before. Now I have a lot of work to do testing through this.

Nonetheless, _this_ PR is still needed, because current state in `devel` forces the global sanitization method when ran through the plugin, with no outlet.

I'm good if you merge this while I continue my testing with your alternative suggestion.

**Review by AlanCoding — APPROVED** [review]

**Comment by AlanCoding** [comment]
I have implemented usage of this toggle in the inventory plugins transition feature branch here:

https://github.com/ansible/awx/commit/60c9f5b5db698d0bc2abe6b30b12ef8c2b31dbfd

This was tested, and confirmed intended functionality.

I will be completing this and passing it off to formal testing as soon as this and corresponding toggles are added for Azure and gce get merged into `devel`.

File: lib/ansible/plugins/inventory/aws_ec2.py [hunk 3]
```diff
@@ -69,6 +69,16 @@
               False in the inventory config file which will allow 403 errors to be gracefully skipped.
           type: bool
           default: True
+        use_legacy_script_group_name_sanitization:
+          description:
+            - By default this plugin is using a general group name sanitization to create safe and usable group names for use in Ansible.
+              This toggle allows those migration from the old ec2.py inventory script that want to continue using the old sanitization to do so.
```
Comment (AlanCoding) [inline lib/ansible/plugins/inventory/aws_ec2.py]: > This toggle allows those migration from the old ec2.py

just not for region names.

Comment (bcoca) [reply lib/ansible/plugins/inventory/aws_ec2.py]: can you point out where region names get added in both script and plugin?

Comment (AlanCoding) [reply lib/ansible/plugins/inventory/aws_ec2.py]: ### Script

adding ami_id, does to_safe, and then push

https://github.com/ansible/ansible/blob/d4bd01c7acaa380b0b011822684e8dbc4d1fb251/contrib/inventory/ec2.py#L983-L984

region, does push_group

https://github.com/ansible/ansible/blob/d4bd01c7acaa380b0b011822684e8dbc4d1fb251/contrib/inventory/ec2.py#L1132

Also see https://github.com/ansible/ansible/pull/53458#issuecomment-470648367, it's hard to prove a negative, there's no substitute for just running it.

I only mentioned this because of the wording of the documentation here. I am planning to go the route s-hertel mentioned.

### Plugin

Region names don't get added in the plugin by default. It is a user-defined entry in `keyed_groups`, as in my above examples.

**Comment by AlanCoding** [comment]
Link https://github.com/ansible/ansible/pull/53524

File: lib/ansible/plugins/inventory/aws_ec2.py (hunk 3 above)
Comment (s-hertel) [reply lib/ansible/plugins/inventory/aws_ec2.py]: possible rewording for clarity:
"This toggle allows migration from the old ec2.py and matches the sanitization of groups when the script's `replace_dash_in_groups` option is set to False. To replicate behavior of `replace_dash_in_groups = True`, you will need to replace hyphens with underscores via the `regex_replace` filter for the affected groups."

Comment (bcoca) [reply lib/ansible/plugins/inventory/aws_ec2.py]: that won't remove auto added groups with - just constructed ones, but good rephrase in general

Comment (bcoca) [reply lib/ansible/plugins/inventory/aws_ec2.py]: updated

Comment (AlanCoding) [reply lib/ansible/plugins/inventory/aws_ec2.py]: does aws_ec2 have any auto added groups? I am not aware of any.

File: changelogs/fragments/allow_bad_things.yml [hunk 4]
```diff
@@ -1,2 +1,4 @@
 minor_changes:
-    - Ebmed an overridable static santiization method into base inventory plugin class to allow individual plugins to override
+    - Embed an overridable static santiization method into base inventory plugin class to allow individual plugins to optionally override
```
Comment (goneri) [inline changelogs/fragments/allow_bad_things.yml]: santiization :-)

File: lib/ansible/plugins/inventory/aws_ec2.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/plugins/inventory/aws_ec2.py]: if none , then i don't understand your complaint, if all are added via compose you already control sanitization

Comment (AlanCoding) [reply lib/ansible/plugins/inventory/aws_ec2.py]: The change you are making here is allowing `-` to be returned. We can replace `-` with `_` with what we control, but we _cannot_ replace `_` with `-` after the plugin has already sanitized them - which is the current state of `devel`.

This toggle does not replicate the ec2.py behavior, because that replaced dashes inconsistently. The simple explanation of the functionality of _this_ toggle is "dashes are not replaced". For compatibility, the user is on their own. The reason this is related to compatibility is that it's theoretically impossible to have compatibility until we get something in `devel` that does not replace the dashes.

Comment (bcoca) [reply lib/ansible/plugins/inventory/aws_ec2.py]: you can still replace them yourself, as @s-hertel shows you above, i'm not sure we want to make a 'toggle per possible path' when it is already possible to control this

Comment (AlanCoding) [reply lib/ansible/plugins/inventory/aws_ec2.py]: yes, that's fine, the toggle per-keyed-group would been a different solution for this, but this PR is sufficient.

**Comment by AlanCoding** [comment]
I've been working at @s-hertel's solution, and it's turned out to be much more complex than I anticipated, because we have to throw in other very uncommon usage forms along with this to get all the compatibility elements.

One problem is that we are running with `ANSIBLE_JINJA2_NATIVE=true` (https://github.com/ansible/ansible/issues/37467) `ANSIBLE_TRANSFORM_INVALID_GROUP_CHARS=false` (https://github.com/ansible/ansible/pull/52748), in addition, we are dealing with some fairly complex data structures.

The issue is a conflict with the jinja2 native setting. If you use the simple `key: tags | regex_replace('-', '_')`, then it combines the entire dictionary into a single group name which is extremely bad, like `tag__value_name__value2_name2__...`. I have a solution that seems to work for `tags` specifically, for now...

What it looks like now:

```yaml
keyed_groups:
  - prefix: tag
    key: tags.keys() | map('regex_replace', "-", "_") | list
    parent_group: tags
  - prefix: tag
    key: dict(tags.keys() | map('regex_replace', "-", "_") | list | zip(tags.values() | map('regex_replace', "-", "_") | list))
    parent_group: tags
```

This seems to get the group names we need. But I also kind of have a problem with a solution that's this complicated, for a problem which isn't all that complicated.

Going with this approach adds a _lot_ of overhead. It's really hard to follow, and I have to bring up https://github.com/ansible/ansible/pull/53446 again. If that was done, then all I would have to do is use "unsafe" in the keyed group construction for zones, and then problem would be solved.

**Comment by bcoca** [comment]
@AlanCoding i find the 'unsafe' solution to be more complex and limited, its also 'hardcoded' , in any case i would prefer to leave it on the specific plugin vs injecting it into constructed.

**Comment by AlanCoding** [comment]
I know that https://github.com/ansible/ansible/pull/53446 doesn't have tests passing, so I re-opened https://github.com/ansible/ansible/pull/51958 and modified it so that it should be passable.

Let me state my _most_ preferred outcome.

I would prefer that this be merged and that https://github.com/ansible/ansible/pull/51958 be merged. This is the same thing I said earlier in this conversation.

I appreciate the alternative solution, but I don't think _my own team_ would approve work that I did that looks like what the solution above is. In other words, I think there's a good chance that solution is un-shippable. If you want to give me another solution which hasn't been discussed yet, then I will be happy to dive in and try it.

**Comment by AlanCoding** [comment]
> use_legacy_script_group_name_sanitization

This does not perform legacy script group sanitization. It does not do it for the case of `replace_dash_in_groups = True`. It does not do it for the case of `replace_dash_in_groups = False`. Pretty soon, the default behavior will not match either of those cases either. All you can honestly say is that it does _a different_ kind of group name sanitization. Replicating what the script does is hard, and we've already established that the plugins will not offer that.

> This option allows you to override that, in efforts to allow migration from the old ec2.py and matches the sanitization of groups when the script's ``replace_dash_in_groups`` option is set to ``False``. To replicate behavior of ``replace_dash_in_groups = True`` with constructed groups, you will need to replace hyphens with underscores via the regex_replace filter for those entries.

Something just doesn't add up about this.

If I am customizing to the point of using `regex_replace`, then I'm re-writing this piece of functionality. I'm just doing it in jinja2 syntax.

So if the user is doing all that work themselves, why are you doing _any_ form of replacement. Your replacements can only mess up the replacements that user is trying to do.

I've taken these observations, and tried to combine with the good work you did here, and I made a 2nd alternative to this PR, up at: https://github.com/ansible/ansible/pull/53589

**Comment by s-hertel** [comment]
@AlanCoding Can you share a gist of how this doesn't match replace_dash_in_groups = False sanitization? I don't have a regex_replace in my gist for `replace_dash_in_groups = False`. Appears to match the sanitization behavior for me.

-1 to #53589 since it's 1) plugin-wide and not host var-specific and 2) now a user has to sanitize all the things themselves, beyond replacing hyphens with underscores. Your proposal for an unsafe option on keyed groups was better IMO since at least it was limited in the things that would be made unsafe (then a user would only need to tread lightly/sanitize that particular group, and it was a known evil, rather than hidden in the mass of things created by the plugin, just waiting for a user to trip over it).

Before removing sanitization altogether and making users figure it out **all** of it if they're trying to replicate script behavior, I'd sooner hide it in code with a toggle as I've mentioned a number of times now. Here's an example that would be mergeable on top of this one, both with toggles for both replace_dash_in_groups and legacy host vars: #53581

**Comment by AlanCoding** [comment]
> Can you share a gist of how this doesn't match `replace_dash_in_groups = False` sanitization? I don't have a regex_replace in my gist for `replace_dash_in_groups = False`.

With https://github.com/ansible/ansible/pull/53544, it will not replace unicode, although the script did replace unicode.

> -1 to #53589 since it's 1) plugin-wide and not host var-specific and 2) now a user has to sanitize all the things themselves, beyond replacing hyphens with underscores. Your proposal for an unsafe option on keyed groups was better IMO since at least it was limited in the things that would be made unsafe (then a user would only need to tread lightly/sanitize that particular group, and it was a known evil, rather than hidden in the mass of things created by the plugin, just waiting for a user to trip over it).

If the user doesn't use `keyed_groups`, this will not affect them. If they wanted group name sanitization, they already have that in the default behavior.

The added toggle in all proposed changes has help text drenched in warnings. It is very frustrating to ask for Ansible to not sanitize group names (the current behavior for most things in the current 2.7 version, ec2 is just one case) and have it still replace anything.

The level of agreement was extremely high that we would handle `regex_replace` on the user side. #53544 is logical consequence of that.

_This_ PR would be fine to address my problem, assuming the same thing was done for the other sources. My concern is that any kind of "legacy" setting will be an extremely volatile contract, because I know just how hard it is to obtain legacy behavior. Heck, I don't even have it right yet, and I'm doing testing every single day.

The only thing I need is a setting for Ansible to take its hands off the wheel long enough for me to use my own customizations to create group names like `us-east-1`. That's a simple request. There is a specific method being called in a single place in code. I need a toggle to tell Ansible "do not call this function".

If https://github.com/ansible/ansible/pull/53589 was merged, the user would have very high confidence that its functionality would not change.

Seeing #53581 confirms all my worst fears about this PR. Calling anything with the inventory plugin "legacy" regarding the script fundamentally breaks the structure of the contract between Ansible core functionality and the compatibility content we have been trying to build out.

**Comment by AlanCoding** [comment]
If we can agree in principle that #53524 works, then that have 👍 from me, it would solve my problem. I didn't think that would be approved because of:

```python
self._sanitize_group_name = lambda name: name
```

in gcp_compute.

If we go that route, I would like to agree that nothing like #53581 will be done.

**Comment by s-hertel** [comment]
Re: https://github.com/ansible/ansible/pull/53458#issuecomment-471528717

Your first point isn't true in the context of this pull request. This pull request introduces a legacy regex for aws_ec2 so it will be replacing unicode like it always has. Did you have other ideas for why this doesn't match `replace_dash_in_groups = False`? It does match `replace_dash_in_groups = False` behavior as far as I see so I'm not sure what the problem is with this pull request.

For your second point, a big point of the plugin system was that users would supply keyed_groups since the plugin doesn't create them. And yes, they can turn the toggle off. But if they're trying to replicate script behavior this looks significantly harder to do. It's going to have the same problem you have with unicode in #53544. And now a user has to populate their config with sanitizing regexes. Now not only do you have hyphens to deal with, but you have all the invalid characters to address.

The only real way to make sure something doesn't change is have a test for it. Once we can settle on a solution #53435 will need to be testing just about every bit of it possible. I'm imagining how we'd test #53589 for compatibility and it doesn't seem like a good idea.

I prefer my pull request less than just using this PR as the solution. It shouldn't be volatile, as it will need to have thorough testing in #53435. I don't see how it breaks a contract. Anyway, I only opened it because it would significantly simplify things for users. I'm not fighting for it, it's just better than some of the alternatives you're proposing where the user is really left to codify their config file to an extreme.

**Comment by s-hertel** [comment]
I'm personally okay with #53524 being merged after this PR. The reason I think it's okay for GCP compute because the script did no sanitization at all. The other scripts did and had a very limited scope of invalid characters put in groups, which is why I wouldn't want to see like this done across the board.

**Comment by AlanCoding** [comment]
Closed https://github.com/ansible/ansible/pull/53524, since that's all included here now, the discussion can be consolidated. I haven't addressed the above points yet (but we are getting on the same page), I have some testing and use cases to chew through.

File: lib/ansible/plugins/inventory/aws_ec2.py [hunk 5]
```diff
@@ -518,9 +533,14 @@ def verify_file(self, path):
         return False
 
     def parse(self, inventory, loader, path, cache=True):
+
         super(InventoryModule, self).parse(inventory, loader, path)
 
-        config_data = self._read_config_data(path)
+        self._read_config_data(path)
+
+        if self.get_option('use_contrib_script_compatible_sanitization:'):
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/aws_ec2.py]:
```suggestion
        if self.get_option('use_contrib_script_compatible_sanitization'):
```

File: lib/ansible/plugins/inventory/gcp_compute.py [hunk 6]
```diff
@@ -325,6 +335,9 @@ def parse(self, inventory, loader, path, cache=True):
         config_data = {}
         config_data = self._read_config_data(path)
 
+        if self.get_option('use_contrib_script_compatible_sanitization:'):
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/gcp_compute.py]:
```suggestion
        if self.get_option('use_contrib_script_compatible_sanitization'):
```

File: lib/ansible/plugins/inventory/azure_rm.py [hunk 7]
```diff
@@ -227,14 +241,18 @@ def parse(self, inventory, loader, path, cache=True):
         super(InventoryModule, self).parse(inventory, loader, path)
 
         self._read_config_data(path)
+
+        if self.get_option('use_contrib_script_compatible_sanitization:'):
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/azure_rm.py]:
```suggestion
        if self.get_option('use_contrib_script_compatible_sanitization'):
```

**Comment by AlanCoding** [comment]
For the case of the old Azure script:

https://github.com/ansible/ansible/blob/b187e5a5c0029befe6fb73daea68bfd5f1a06625/contrib/inventory/azure_rm.py

I can offer no guarantees that it consistently used the name sanitization method when creating groups. `ec2.py` did not. I have not yet done combined testing of this PR + Azure infrastructure + my jinja2 dynamic code generation code.

For the case of `ec2.py`, I have found with only a very basic level of testing that regions and zones bypassed the name sanitization method. Are other categories of groups also saved in their non-sanitized form? I can't guarantee.

This is why #53589 exists. If you do a simpler implementation, then it enables me to resolve issues that might pop up. This more user-restrictive solution does now allow for that. I understand that you have a _strong_ preference for being more user-restrictive. That is your choice to make.

**Comment by s-hertel** [comment]
Regions and zones in ec2 contain hyphens. They don't contain other invalid characters, and this is why it is still consistent with replace_dash_in_groups = False. The other groups (I am not talking about route53, elasticache and RDS right now as I've been informed tower isn't interested in these features of ec2 currently and they will have their own plugins), may contain hyphens or not (depending on replace_dash_in_groups), but will not contain other unsafe characters. This is consistent with ec2.py's `replace_dash_in_groups = False`.

I don't currently have an Azure account to test rigorously, but it looks like the only group not being sanitized is location. What kind of characters can location contain? If it contains hyphens but not other invalid characters then it will also be compatible with the azure script's `replace_dash_in_groups = False` behavior.

**Comment by AlanCoding** [comment]
Location syntax (as hostvars) for `azure_rm.py` looks like `eastus2`. That is just one value, but it sets the tone, they _probably_ don't have any non-ascii symbols at all.

**Comment by AlanCoding** [comment]
Local CLI testing with cloud accounts:

heuristics being looked for:
 - aws
   - dashes in region names
   - dashes in group names formed from tags
 - azure
   - dashes in security / resource group group names
 - gcp_compute
   - periods in groups formed from IP addresses
   - dashes in groups formed from entries in tags

Obtained by

 - `use_contrib_script_compatible_sanitization: true` in inventory config
 - `ANSIBLE_TRANSFORM_INVALID_GROUP_CHARS=false` before command

Confirmed functionality manual testing for above points.

**Review by AlanCoding — APPROVED** [review]

**Comment by AlanCoding** [comment]
I also confirmed that this does not break the _tower_ inventory plugin

```
$ ansible-inventory -i private/tower/tower.yml --list --export -vvv
ansible-inventory 2.8.0.dev0
  config file = None
  configured module search path = [u'/Users/alancoding/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/alancoding/Documents/repos/ansible/lib/ansible
  executable location = /Users/alancoding/.virtualenvs/ansible/bin/ansible-inventory
  python version = 2.7.11 (default, Oct 17 2016, 14:59:40) [GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
No config file found; using defaults
host_list declined parsing /Users/alancoding/Documents/repos/ansible-inventory-file-examples/private/tower/tower.yml as it did not pass it's verify_file() method
script declined parsing /Users/alancoding/Documents/repos/ansible-inventory-file-examples/private/tower/tower.yml as it did not pass it's verify_file() method
[DEPRECATION WARNING]: Ignoring invalid character(s) "set([u'-'])" in group name (dashy-dash). This feature will be removed in version 2.12. Deprecation warnings can be disabled by
 setting deprecation_warnings=False in ansible.cfg.
Parsed /Users/alancoding/Documents/repos/ansible-inventory-file-examples/private/tower/tower.yml inventory source with auto plugin
{
    "_meta": {
        "hostvars": {
            "aewfefe": {
                "remote_tower_enabled": "true", 
                "remote_tower_id": 16774
            }
        }
    }, 
    "all": {
        "children": [
            "dashy-dash", 
            "ungrouped"
        ]
    }, 
    "dashy-dash": {
        "hosts": [
            "aewfefe"
        ]
    }
}
```

This is existing behavior, and this is what I want, please don't change it.
